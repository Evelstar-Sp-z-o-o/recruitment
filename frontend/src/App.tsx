import { FC } from 'react';

import Root from '@/src/pages/Root';
import AppProviders from '@/src/utils/providers/AppProviders';

import './styles/main.scss';

const App: FC = () => {
  return (
    <AppProviders>
      <Root />
    </AppProviders>
  );
};

export default App;
