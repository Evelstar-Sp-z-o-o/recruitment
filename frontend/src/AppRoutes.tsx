import { Route, Routes, useLocation } from 'react-router-dom';

import Sidebar from './components/Sidebar';
import MainLayout from './components/layout/MainLayout';
import PostsLayout from './components/layout/PostsLayout';
import CreatePost from './pages/CreatePost';
import Home from './pages/Home';
import UpdatePost from './pages/UpdatePost';

const AppRoutes = () => {
  const location = useLocation();
  const background = location.state && location.state.background;

  return (
    <div className="main">
      <Sidebar />
      <Routes location={background || location}>
        <Route path="/" element={<MainLayout />}>
          <Route index element={<Home />} />
          <Route path="/posts" element={<PostsLayout />}>
            <Route path="update/:postId" element={<UpdatePost />} />
          </Route>
        </Route>
      </Routes>
      {background && (
        <Routes>
          <Route path="create" element={<CreatePost />} />
        </Routes>
      )}
    </div>
  );
};

export default AppRoutes;
