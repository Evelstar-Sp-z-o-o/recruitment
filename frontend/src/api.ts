import type { Post } from './types';

export const BASE_API_URL = 'http://localhost:3000/api';

const getPosts = async () => {
  const ENDPOINT = '/posts';

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Posts list.');
  }

  const data = await res.json();

  return data as unknown as Post[];
};

const getSinglePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Post details.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const deletePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'DELETE',
  });

  if (!res.ok) {
    throw new Error('Something went wrong when deleting post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const updatePost = async (postId: string, postData: Partial<Post>) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when updating post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const createPost = async (postData: Partial<Post>) => {
  const ENDPOINT = `/posts`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when creating post.');
  }

  const data = await res.json();

  return data as unknown as Post;
};

export { getPosts, getSinglePost, updatePost, createPost, deletePost };
