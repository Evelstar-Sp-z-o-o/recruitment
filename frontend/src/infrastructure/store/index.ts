import { useDispatch, useSelector } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';

import apiService from './services/apiService';
import userSlice from './slices/user';

const store = configureStore({
  reducer: {
    [apiService.reducerPath]: apiService.reducer,
    user: userSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiService.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector = useSelector.withTypes<RootState>();

export default store;
