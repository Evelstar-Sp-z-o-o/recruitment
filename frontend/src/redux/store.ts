import { configureStore } from '@reduxjs/toolkit';

import responseModalReducer from './responseModalSlice';

const store = configureStore({
  reducer: {
    responseModal: responseModalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
