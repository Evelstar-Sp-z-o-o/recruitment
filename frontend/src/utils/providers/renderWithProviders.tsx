import { JSX, PropsWithChildren } from 'react';

import { render } from '@testing-library/react';

import AppProviders from './AppProviders.tsx';

export const renderWithProviders = (children: PropsWithChildren<JSX.Element>) => {
  return render(<AppProviders>{children}</AppProviders>);
};
