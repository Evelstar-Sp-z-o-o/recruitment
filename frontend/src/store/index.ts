import { postsApi } from '@/src/store/api/posts';
import { configureStore, createSlice } from '@reduxjs/toolkit';

const initialUser: null | string = null;
const initialLogin = false;

const userSlice = createSlice({
  name: 'user',
  initialState: initialUser,
  reducers: {
    setUser(_state, action) {
      return action.payload;
    },
  },
});

const loginSlice = createSlice({
  name: 'login',
  initialState: initialLogin,
  reducers: {
    setLogin(_state, action) {
      return action.payload;
    },
  },
});

export const { setUser } = userSlice.actions;
export const { setLogin } = loginSlice.actions;

export * from './api/posts.ts';

export const store = configureStore({
  reducer: {
    [postsApi.reducerPath]: postsApi.reducer,
    user: userSlice.reducer,
    login: loginSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(postsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
