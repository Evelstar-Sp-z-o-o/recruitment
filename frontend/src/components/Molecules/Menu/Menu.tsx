import { FC, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';

import Unknown from '@/src/assets/noUserIcon.svg';
import Avt from '@/src/assets/userIcon.svg';
import AddNewPost from '@/src/components/Atoms/AddNewPost/AddNewPost';
import MenuLinkItem from '@/src/components/Atoms/MenuLinkItem/MenuLinkItem';
import MenuLogin from '@/src/components/Atoms/MenuLogin/MenuLogin';
import LoginModal from '@/src/components/Molecules/LoginModal/LoginModal';
import PostModal from '@/src/components/Molecules/PostModal/PostModal';
import { setLogin, setUser, RootState } from '@/src/store';
import Avatar from '@mui/material/Avatar';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Drawer from '@mui/material/Drawer';
import MenuList from '@mui/material/MenuList';
import Typography from '@mui/material/Typography';

interface IMenuProps {
  open: boolean;
  toggleMenu: () => void;
  handleResponse: () => void;
  postsCount: number;
}

const Menu: FC<IMenuProps> = ({ open, toggleMenu, handleResponse, postsCount }) => {
  const dispatch = useDispatch();
  const { t } = useTranslation();
  const user: string = useSelector<RootState>((state) => state.user);
  const [openLogin, setOpenLogin] = useState<boolean>(false);
  const [openPostModal, setOpenPostModal] = useState<boolean>(false);

  const handleToggleModal = () => {
    setOpenLogin((prevState) => !prevState);
  };

  const handleTogglePostModal = () => {
    setOpenPostModal((prevState) => !prevState);
  };

  const handleSnackbar = (isOpen) => {
    dispatch(setLogin(isOpen));
  };

  const handleLogout = () => {
    toggleMenu(false);
    handleSnackbar(true);
    localStorage.setItem('wasHomeModalShown', true);
    sessionStorage.removeItem('user');
    dispatch(setUser(null));
  };

  return (
    <Drawer component="aside" open={open} onClose={toggleMenu}>
      <LoginModal isOpen={openLogin} handleClose={handleToggleModal} handleSnackbar={handleSnackbar} />
      <Box className="sideMenu">
        <Container className="menuHeader" maxWidth="xl">
          <Avatar alt="User" src={user ? Avt : (Unknown as string)} />
          <AddNewPost onClick={handleTogglePostModal} />
        </Container>
        <Container maxWidth="xl" sx={{ mb: '1rem' }}>
          <Typography color="text.secondary" sx={{ fontWeight: '700' }}>
            {user}
          </Typography>
        </Container>

        <Box component="nav">
          <MenuList>
            <MenuLinkItem target="/" label={t('menu.home')} />
            {user ? <MenuLinkItem target="/profile" label={t('menu.profile')} /> : null}
            {user ? (
              <MenuLogin handleLogout={handleLogout} />
            ) : (
              <MenuLogin isLogin handleToggleModal={handleToggleModal} />
            )}
          </MenuList>
        </Box>
        <PostModal
          open={openPostModal}
          close={handleTogglePostModal}
          response={handleResponse}
          postsCount={postsCount}
        />
      </Box>
    </Drawer>
  );
};

export default Menu;
