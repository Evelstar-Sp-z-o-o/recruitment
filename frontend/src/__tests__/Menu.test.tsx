import React from 'react';

import Menu from '@/src/components/Molecules/Menu/Menu';
import { store } from '@/src/store';
import { setUser } from '@/src/store';
import { renderWithProviders } from '@/src/utils/providers/renderWithProviders';
import '@testing-library/jest-dom';
import { screen, fireEvent, waitFor } from '@testing-library/react';

describe('Menu', () => {
  const handleResponse = vi.fn();
  const toggleMenu = vi.fn();

  vi.mock('react-i18next', async (importOriginal) => {
    const actual = await importOriginal();
    return {
      ...actual,
      useTranslation: () => ({
        t: (key: string) => key,
      }),
      I18nextProvider: ({ children }) => <div>{children}</div>,
    };
  });

  it('renders correctly when open', () => {
    renderWithProviders(<Menu open={true} toggleMenu={toggleMenu} handleResponse={handleResponse} />);

    expect(screen.getByRole('button', { name: /addPost/i })).toBeInTheDocument();
    expect(screen.getByRole('img', { name: /User/i })).toBeInTheDocument();
  });

  it('opens and closes the login modal correctly', async () => {
    renderWithProviders(<Menu open={true} toggleMenu={toggleMenu} handleResponse={handleResponse} />);

    fireEvent.click(screen.getByRole('menuitem', { name: /menu.log.in/i }));

    await waitFor(() => {
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByTestId('CloseIcon'));

    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  it('displays user email when logged in and shows the correct menu items', () => {
    store.dispatch(setUser('test@example.com'));

    renderWithProviders(<Menu open={true} toggleMenu={toggleMenu} handleResponse={handleResponse} />);

    expect(screen.getByText('test@example.com')).toBeInTheDocument();
    expect(screen.getByText('menu.profile')).toBeInTheDocument();
    expect(screen.queryByText('menu.home')).toBeInTheDocument();
    fireEvent.click(screen.getByRole('menuitem', { name: /menu.log.out/i }));
  });

  it('calls handleLogout and updates store on logout', () => {
    store.dispatch(setUser('test@example.com'));

    renderWithProviders(<Menu open={true} toggleMenu={toggleMenu} handleResponse={handleResponse} />);

    fireEvent.click(screen.getByRole('menuitem', { name: /menu.log.out/i }));

    expect(toggleMenu).toHaveBeenCalledWith(false);

    const state = store.getState();
    expect(state.user).toBeNull();
  });

  it('should match snapshot', () => {
    const { asFragment } = renderWithProviders(
      <Menu open={true} toggleMenu={toggleMenu} handleResponse={handleResponse} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
