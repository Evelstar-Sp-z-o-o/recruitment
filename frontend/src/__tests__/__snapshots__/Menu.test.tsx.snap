// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > should match snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
